API: Create Preliminary DTypeMeta class and np.dtype subclasses

seberg opened the pull request:
This makes all dtypes being subclasses of a:

```
class DType(np.dtype):
    is_abstract = False
    is_flexible = False  % True for void, strings, and datetimes
    singleton = prototype_dtype_instance
    # other information defined on the singleton currently
```

subclass of np.dtype.

Right now this happens by replacing the Type field of of the
singleton instance. This is valid for types/objects not created
in python (not HeapTypes). It is necessary to have this functionality
for continuing support of user defined legacy dtypes.

---

This is basically the draft start for the proposed NEP 42. I would not want to expose much more, if we have some agreement that `np.dtype[np.float64]` is a nice API, I would be happy to add such small things (with a `PreliminaryDTypeAPIWarning`).

My hope is this:

1. We can agree to merge something like this, even if it may still be a bit in flux.
2. This enables other smaller changes, which are somewhat unrelated. E.g. use new methods stored on these DTypes to:
   * reimplement the current `np.array(...)` coercion. 
   * Reimplement how sorting is handled (add new sort functions) if someone cares for it, it is low priority for me.
   * Reimplement how we organize casting.

Most of which has to be defined in more technical detail and with specific decisions and would be part of NEP 42 though.

*This PR aims to add **no** new public API with the exception of barely useful DType classes!*. In aprticularly *no* C-API is added.

---

As of now this is probably still a bit draft. The whole thing is a bit ugly, but I think it is actually somewhat fine, and we have to do this "heap allocated static type" for things such as `quaternions`.

seberg committed:
API: Create Preliminary DTypeMeta class and np.dtype subclasses

This makes all dtypes being subclasses of a:

class DType(np.dtype):
    is_abstract = False
    is_flexible = False  % True for void, strings, and datetimes
    singleton = prototype_dtype_instance
    # other information defined on the singleton currently

subclass of np.dtype.

Right now this happens by replacing the Type field of of the
singleton instance. This is valid for types/objects not created
in python (not HeapTypes). It is necessary to have this functionality
for continuing support of user defined legacy dtypes.

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -0,0 +1,245 @@
+        {"type",
+                T_OBJECT, offsetof(PyArray_DTypeMeta, scalar_type), READONLY, NULL},
+        {"_flexible",
+                T_BYTE, offsetof(PyArray_DTypeMeta, is_flexible), READONLY, NULL},
```
Hmmm, seems these do not actually kick in right now. They could probably be removed from this PR anyway, but have to explore why `type(np.dtype(np.float64))._flexible` fails

seberg replied:
Seems to work now (or maybe I just tried it incorrectly for a while); Although it is not visible as such, but that is similar to python ABCs.

seberg commented:
The only thing from python that this is supposed to change, is that:
```
type(np.float64)
```
should *not* be `np.dtype`, but instead gives/prints: `numpy.dtype[numpy.float64]`. I do not really care right now how we want to name this in the future. Although I actually like the square bracket idea. Possibly even to get the class. Since that removes the need for putting a second set of names somewhere if we need `np.dtype[np.inexact]`.

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -0,0 +1,245 @@
+        .tp_flags = Py_TPFLAGS_DEFAULT | Py_TPFLAGS_HAVE_GC,
+        .tp_doc = "Preliminary NumPy API: The Type of NumPy DTypes (metaclass)",
+        .tp_members = dtypemeta_members,
+        .tp_base = &PyType_Type,
```
Ah, this (and similar in `np.dtype`) is not strictly valid, and I have to do that during module loading. Maybe the reason for the above not working...

seberg committed:
Possible initialization safety fixups and fix API hash.

(currently no public API changes, but the hash changed).
The type safety may not be important as such, but there used
to be issues, at least with MSVC about this, and clang warns.

seberg commented on numpy/core/code_generators/genapi.py:
```diff
@@ -325,9 +328,19 @@ def array_api_define(self):
+            f"extern NPY_NO_EXPORT {self.internal_type} {mangled_name};\n"
+            # And define the name as: (*(type *)(&mangled_name))
+            f"#define {self.name} (*({self.ptr_cast} *)(&{mangled_name}))\n"
+        )
```
The API generation changes are the ugliest part here... Unfortunately, I really want to hide everything away (unless it is an internal build), and have the `PyArrayDescr_Type` defined as `PyTypeObject` even internally. But I also need to initialize/allocate it statically as a `PyArray_DTypeMeta`.

seberg committed:
Small fixups/comments

mattip commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -0,0 +1,237 @@
+     * Initialize the struct fields similar to static code:
+     */
+    /* Copy in np.dtype since it shares most things */
+    memcpy(dtype_class, &PyArrayDescr_Type, sizeof(PyArray_DTypeMeta));
```
You shouldn't copy from the start of the object. It overrides the PyObject fields, which for PyPy holds the link to the PyPy object. I am pretty sure this is what is causing the segfaults on PyPy. So something like
```
memcpy((char*)dtype_class + sizeof(PyObject), (char*)&PyArrayDescr_Type + sizeof(PyObject), sizeof(PyArray_DTypeMeta) - sizeof(PyObject));
```

I also had this [problem](https://github.com/numpy/numpy/pull/12585/files#diff-f3387afa660b73f8f37f0978193799d9R4442)

seberg replied:
Ah then a `PyObject_Init()` on the type should be the right thing, lets see (I was not able to reproduce locally, strangely enough, even with the pypy nightly).

seberg replied:
OK, after a lot of pain (and still not being quite sure where things go wrong, but maybe subtly different things at different times), it must indeed be the incorrect initialization of the `Py(Var)Object` fields. This now seems to work fine, by creating a new static "prototype" which is copied. That ensures that the code indeed looks identical to a static type declaration, instead of approximating it in a way that was not quite good enough for PyPy.

seberg committed:
Create a static prototype to copy, so that it is more obvious what the intention is

seberg committed:
STY: Just small indentation fixup

eric-wieser commented:
We've had a lot of confusion in past discussions about `issubclass` vs `isinstance` relationships. Can you add some tests to this PR to make it easy to tell at a glance what is actually implemented here?

seberg committed:
Add tests and fixup __name__

seberg commented:
Thanks for having a look Eric. I had put this mainly in the hope to get quick judgment as to this "static" type being a bit weird. But by now I think it is a good first step (even if not pretty). Of course we could at some point decide to make all dtypes heaptypes, but that is fully changable.

There was a bug with `__name__`, because the name part of the type (for a static type) cannot include `.`, that is fixed now. The test is minimal, but it does test pretty much everything that is actually changed.

seberg committed:
Use "parametric" instead of "flexible" and add tests for attributes

Note that these attributes are *not* fixed yet, just a start to
have something.

seberg committed:
Improve docs and test slightl; Reject also kwargs in dtype from type creation

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -0,0 +1,268 @@
+legacy_dtype_default_new(PyArray_DTypeMeta *self,
+        PyObject *args, PyObject *kwargs)
+{
+    /* TODO: This should allow endianess and possibly metadata */
```
I am happy to just disable this completely for now as well. Or, put a PreliminaryAPI warning here. Which I think would be nice in general, but I am not sure I care too much about right away (it is not like this is a tricky API choice), that `type(instance)()` gives the default instance.

The main annoyance about this PR is, how to create the TypeStruct in a way that is hidden from the public.

rgommers replied:
Warnings are probably annoying, given that you'll then have to filter them out again if you want to use this internally (applies to any preliminary API). How about just putting a big warning in the docs for any such object? And on the Python side, keep it out of the main namespace.

seberg replied:
True, as this is the python side, for a long time it probably only matters in tests though. `np.dtype` does not really have a test right now. I could change the repr away from `np.dtype[float64]` if you prefer. E.g. just add an underscore to make it "private".

seberg committed:
PyDict_GET_SIZE is not public (or at least not on older python)

seberg committed:
small fixups/name changes

seberg commented:
@eric-wieser if you have a larger chunk of concentration cycles (I know those are hard to get by, and its not getting easier right now likely)... I am not sure if you looked at this before much. I do see most of this a bit in flux (e.g. as soon as we tag on more to the DType class, it may be necessary to create them explicitly).

But I would be happy if we can get something in this direction going shortly after merging the NEP, which I hope we can manage in less then two weeks and I think you would be the best reviewer, or to suggest larger changes.

mattip commented:
Now that 1.19 is out, we should move this forward

seberg commented:
Yes, that would be great. I am not sure there is much for me to do here right now though. It requires some review. If anyone wants to talk about it or so, I am happy to do that.

mattip commented:
LGTM. It may need some iteration and adjustment as the work progresses, but I think the approach is solid. Since this unlocks more work I would like to put it in soonish. @eric-wieser  thoughts?

mattip commented:
Thanks @seberg. This may need some fine tuning but let's see how it goes.

tacaswell commented:
This has broken compilation with the cython + cpython master.  

You get failures like:

```
    C compiler: gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC

    compile options: '-DNPY_INTERNAL_BUILD=1 -DHAVE_NPY_CONFIG_H=1 -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE=1 -D_LARGEFILE64_SOURCE=1 -Inumpy/core/include -Ibuild/src.linux-x86_64-3.10/numpy/core/include/numpy -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/home/tcaswell/.virtualenvs/bleeding/include -I/home/tcaswell/.pybuild/bleeding/include/python3.10 -Ibuild/src.linux-x86_64-3.10/numpy/core/src/common -Ibuild/src.linux-x86_64-3.10/numpy/core/src/npymath -c'
    extra options: '-std=c99'
    gcc -pthread -shared build/temp.linux-x86_64-3.10/build/src.linux-x86_64-3.10/numpy/core/src/multiarray/_multiarray_tests.o build/temp.linux-x86_64-3.10/numpy/core/src/common/mem_overlap.o -Lbuild/temp.linux-x86_64-3.10 -lnpymath -o build/lib.linux-x86_64-3.10/numpy/core/_multiarray_tests.cpython-310-x86_64-linux-gnu.so
    /usr/bin/ld: build/temp.linux-x86_64-3.10/numpy/core/src/common/mem_overlap.o:/tmp/pip-req-build-cxqd0fcv/numpy/core/include/numpy/ndarraytypes.h:1824: multiple definition of `PyArrayDTypeMeta_Type'; build/temp.linux-x86_64-3.10/build/src.linux-x86_64-3.10/numpy/core/src/multiarray/_multiarray_tests.o:/tmp/pip-req-build-cxqd0fcv/numpy/core/include/numpy/ndarraytypes.h:1824: first defined here
    collect2: error: ld returned 1 exit status
    error: Command "gcc -pthread -shared build/temp.linux-x86_64-3.10/build/src.linux-x86_64-3.10/numpy/core/src/multiarray/_multiarray_tests.o build/temp.linux-x86_64-3.10/numpy/core/src/common/mem_overlap.o -Lbuild/temp.linux-x86_64-3.10 -lnpymath -o build/lib.linux-x86_64-3.10/numpy/core/_multiarray_tests.cpython-310-x86_64-linux-gnu.so" failed with exit status 1
    Running setup.py install for numpy ... error
ERROR: Command errored out with exit status 1: /home/tcaswell/.virtualenvs/bleeding/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-cxqd0fcv/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-cxqd0fcv/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-8ridma62/install-record.txt --single-version-externally-managed --compile --install-headers /home/tcaswell/.virtualenvs/bleeding/include/site/python3.10/numpy Check the logs for full command output.
Exception information:
Traceback (most recent call last):
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/req/req_install.py", line 827, in install
    success = install_legacy(
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/operations/install/legacy.py", line 86, in install
    raise LegacyInstallFailure
pip._internal.operations.install.legacy.LegacyInstallFailure

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/cli/base_command.py", line 188, in _main
    status = self.run(options, args)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/cli/req_command.py", line 185, in wrapper
    return func(self, options, args)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/commands/install.py", line 398, in run
    installed = install_given_reqs(
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/req/__init__.py", line 67, in install_given_reqs
    requirement.install(
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/req/req_install.py", line 845, in install
    six.reraise(*exc.parent)
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_vendor/six.py", line 703, in reraise
    raise value
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/operations/install/legacy.py", line 74, in install
    runner(
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/utils/subprocess.py", line 270, in runner
    call_subprocess(
  File "/home/tcaswell/.virtualenvs/bleeding/lib/python3.10/site-packages/pip/_internal/utils/subprocess.py", line 241, in call_subprocess
    raise InstallationError(exc_msg)
pip._internal.exceptions.InstallationError: Command errored out with exit status 1: /home/tcaswell/.virtualenvs/bleeding/bin/python3 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-cxqd0fcv/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-cxqd0fcv/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-8ridma62/install-record.txt --single-version-externally-managed --compile --install-headers /home/tcaswell/.virtualenvs/bleeding/include/site/python3.10/numpy Check the logs for full command output.
Removed build tracker: '/tmp/pip-req-tracker-76reoduh'
```

I am testing with cpython master which is a bit annoying to bisect given the change to `Py_SIZE` and `Py_TYPE` so I have not chased it down further yet.

seberg commented:
You are running into https://github.com/numpy/numpy/issues/16419, have not figured out yet what I have to move around. I guess those declaration just must not be in that file due to the nature of how it is used, but not exactly sure...

tacaswell commented:
Ah, great (well, from the point of view that I don't have to bisect this further...not sure how long it would have taken me to get to "maybe it is gcc"....)!

anirudh2290 reviewed (commented):
I was just reading through the PR to understand some of the dtype changes. Overall this is a solid PR :) ! Added a few comments, feel free to ignore.

seberg commented on numpy/core/tests/test_dtype.py:
```diff
@@ -1091,6 +1091,40 @@ class dt(np.void):
+        # which are more than just storage information, these would need to be
+        # given when creating a dtype:
+        parametric = (np.void, np.str_, np.bytes_, np.datetime64, np.timedelta64)
+        if dtype.type not in parametric:
```
I have to figure this out before 1.20. I currently see the dtype type as non-existing, i.e. its a class attribute, which the instance simply looks up on the class.  But: I think there may be use-cases which `DType.type` is `NULL`, and in that case, there may be some use in `dtype.type` being non-null.  Although I suspect that is useless in practice (at least unless I expand some of the classmethods to also work as instance-methods in the NEP).

For now, I would view it as being a bad implementation of:
```
class DType:
    type = something
    @property 
    def dtype(self):
        return self.type  # returns the class attribute always.
```
but we store it also on the `dtype` for backward compatibility reasons.

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -0,0 +1,269 @@
+        return -1;
+    }
+
+    snprintf(tp_name, name_length, "numpy.dtype[%s]", scalar_name);
```
True... I suppose the metaclass could probably have a `__repr__` slot to fix some of it. Although, it may be that to actually make it copy-pastable, we need a `.` in there, which may mean it can only be done by making everything HeapTypes (which is not totally unreasonable).

anirudh2290 replied:
Can we do something like numpy.dtype_%s ? Although this may be a bit ugly and little less self-explanatory of it being a subclass of np.dtype.

seberg replied:
We could do various things I guess, but it will be a while before we should seriously consider exposing these in the API, and I think until then, it will be fine to modify the `__repr__` as we like.

seberg commented on numpy/core/src/multiarray/dtypemeta.c:
```diff
@@ -0,0 +1,269 @@
+     * in the future when we implement HeapTypes (python/dynamically
+     * defined types). It should be revised at that time.
+     */
+    assert(0);
```
Thanks for that and the other comments.  This one was intentional, because I thought if we actually start using it, we should check that it actually works correctly :).
